lib/Parse/ParseGeneric.cpp

```diff
@@ -261,22 +262,46 @@ ParserStatus Parser::parseGenericWhereClause(
     if (Tok.is(tok::colon)) {
       // A conformance-requirement.
       SourceLoc ColonLoc = consumeToken();
-
-      // Parse the protocol or composition.
-      ParserResult<TypeRepr> Protocol = parseTypeForInheritance(
-          diag::expected_identifier_for_type,
-          diag::expected_ident_type_in_inheritance);
-      
-      if (Protocol.isNull()) {
-        Status.setIsParseError();
-        if (Protocol.hasCodeCompletion())
-          Status.setHasCodeCompletion();
-        break;
+      Identifier TypeId;
+
+      if (Tok.is(tok::identifier) &&
+          getLayoutConstraintInfo(TypeId = Context.getIdentifier(Tok.getText()),
+                                  Context)) {
+        // Parse a layout constraint.
+        auto LayoutName = TypeId;
+        auto LayoutLoc = consumeToken();
+        auto LayoutInfo = parseLayoutConstraint(LayoutName);
+        if (!LayoutInfo.isKnownLayout()) {
+          // There was a bug in the layout constraint.
+          Status.setIsParseError();
```

Comment:
It is expected. The input may be syntactically wrong and which case we would get the UnknownLayoutConstraint back.